File: java/src/main/java/com/gushenge/atools/ui/ArcButton.java
package com.gushenge.atools.ui;


import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.StateListDrawable;
import android.graphics.drawable.shapes.RoundRectShape;
import android.util.AttributeSet;
import android.view.Gravity;
import android.widget.Button;
import android.widget.TextView;

/**
 * Created by Gushenge on 2019/06/14.
 * <p>
 * email : cikexiaosi@gmail.com
 * <p>
 * https://github.com/Gushenge
 */


@SuppressLint("AppCompatCustomView")
public class ArcButton extends TextView {
    public static int[]	mNormalState		= new int[] {};
    public static int[]	mPressState			= new int[] { android.R.attr.state_pressed, android.R.attr.state_enabled };
    private int			mRadius				= 1000000000;																			//默认的圆角半径

    //默认文字和背景颜色
    private int			mBgNormalColor		= Color.GRAY;
    private int			mBgPressedColor		= Color.DKGRAY;
    private int			mTextNormalColor	= Color.BLACK;
    private int			mTextPressedColor	= Color.BLACK;

    public ArcButton(Context context){
        super(context);
        initUI();
    }
    public ArcButton(Context context, AttributeSet src){
        super(context,src);
        initUI();
    }

    private void initUI(){
        setGravity(Gravity.CENTER);
        buildDraweableState();
        buildColorDrawableState();
    }

    /**
     * 构建图片drawble
     */
    private void buildColorDrawableState(){
        ColorStateList colorStateList = new ColorStateList(new int[][] { mPressState, mNormalState },
                new int[] { mTextPressedColor, mTextNormalColor });
        setTextColor(colorStateList);
    }

    /**
     * 构建背景Drawble
     */
    private void buildDraweableState(){

        float outRectr[] = new float[] { mRadius, mRadius, mRadius, mRadius, mRadius, mRadius, mRadius, mRadius };
        //创建状态管理器
        StateListDrawable drawable = new StateListDrawable();
        /**
         * 注意StateListDrawable的构造方法我们这里使用的
         * 是第一参数它是一个float的数组保存的是圆角的半径，它是按照top-left顺时针保存的八个值
         */
        //创建圆弧形状
        RoundRectShape rectShape = new RoundRectShape(outRectr, null, null);
        //创建drawable
        ShapeDrawable pressedDrawable = new ShapeDrawable(rectShape);
        //设置我们按钮背景的颜色
        pressedDrawable.getPaint().setColor(mBgPressedColor);
        //添加到状态管理里面
        drawable.addState(mPressState, pressedDrawable);

        //		ShapeDrawable disableDrawable = new ShapeDrawable(rectShape);
        //		disableDrawable.getPaint().setColor(prssedClor);
        //		disableDrawable.getPaint().setAlpha(125);
        //		drawable.addState(mDisableState, disableDrawable);

        ShapeDrawable normalDrawable = new ShapeDrawable(rectShape);
        normalDrawable.getPaint().setColor(mBgNormalColor);
        drawable.addState(mNormalState, normalDrawable);
        //设置我们的背景，就是xml里面的selector
        setBackgroundDrawable(drawable);
    }

    /**
     * 设置圆角矩形
     *
     * @param radius
     */
    public void setRadius(int radius){
        this.mRadius = radius;
        buildDraweableState();
    }

    /**
     * 设置按钮背景颜色
     *
     * @param normalColor
     * @param prssedClor
     */
    public void setBackgroundColor(int normalColor, int pressedColor){

        mBgNormalColor = normalColor;
        mBgPressedColor = pressedColor;
        buildDraweableState();

    }

    /**
     * 设置按钮文字颜色
     *
     * @param normalColor
     * @param pressedColor
     */
    public void setTextColor(int normalColor, int pressedColor){
        mTextPressedColor = pressedColor;
        mTextNormalColor = normalColor;
        buildColorDrawableState();

    }
    /**
     * 设置按钮背景颜色
     *
     * @param color
     */
    public void setBackgroundColor(int color) {

        mBgNormalColor = color;
        mBgPressedColor = color;
        buildDraweableState();

    }

    /**
     * 设置按钮文字颜色
     *
     * @param color
     */
    public void setTextColor(int color){
        mTextPressedColor = color;
        mTextNormalColor = color;
        buildColorDrawableState();

    }
}

File: kotlin/src/main/java/com/gushenge/atools/ui/ArcButton.kt
package com.gushenge.atools.ui

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.StateListDrawable
import android.graphics.drawable.shapes.RoundRectShape
import android.util.AttributeSet
import android.view.Gravity
import android.view.ViewManager
import android.widget.Button
import android.widget.TextView
import org.jetbrains.anko.custom.ankoView


class ArcButton : TextView {

    constructor(context: Context):super(context)
    constructor(context: Context,attributes: AttributeSet):super(context,attributes)

    //默认的圆角半径
    private var radius = 1000000000

    //默认文字和背景颜色
    private var mBgNormalColor = Color.GRAY
    private var mBgPressedColor = Color.DKGRAY
    private var mTextNormalColor = Color.BLACK
    private var mTextPressedColor = Color.BLACK

    init {
        initUI()
    }

    private fun initUI() {
        gravity = Gravity.CENTER
        buildDraweableState()
        buildColorDrawableState()
    }

    /**
     * 构建图片drawble
     */
    private fun buildColorDrawableState() {
        val colorStateList = ColorStateList(
            arrayOf(mPressState, mNormalState),
            intArrayOf(mTextPressedColor, mTextNormalColor)
        )
        setTextColor(colorStateList)
    }

    /**
     * 构建背景Drawble
     */
    private fun buildDraweableState() {

        val outRectr = floatArrayOf(
            radius.toFloat(),
            radius.toFloat(),
            radius.toFloat(),
            radius.toFloat(),
            radius.toFloat(),
            radius.toFloat(),
            radius.toFloat(),
            radius.toFloat()
        )
        //创建状态管理器
        val drawable = StateListDrawable()
        /**
         * 注意StateListDrawable的构造方法我们这里使用的
         * 是第一参数它是一个float的数组保存的是圆角的半径，它是按照top-left顺时针保存的八个值
         */
        //创建圆弧形状
        val rectShape = RoundRectShape(outRectr, null, null)
        //创建drawable
        val pressedDrawable = ShapeDrawable(rectShape)
        //设置我们按钮背景的颜色
        pressedDrawable.paint.color = mBgPressedColor
        //添加到状态管理里面
        drawable.addState(mPressState, pressedDrawable)

        //		ShapeDrawable disableDrawable = new ShapeDrawable(rectShape);
        //		disableDrawable.getPaint().setColor(prssedClor);
        //		disableDrawable.getPaint().setAlpha(125);
        //		drawable.addState(mDisableState, disableDrawable);

        val normalDrawable = ShapeDrawable(rectShape)
        normalDrawable.paint.color = mBgNormalColor
        drawable.addState(mNormalState, normalDrawable)
        //设置我们的背景，就是xml里面的selector
        background = drawable
    }

    /**
     * 设置圆角矩形
     *
     * @param radius
     */
    fun setRadius(radius: Int) {
        this.radius = radius
        buildDraweableState()
    }
    /**
     * 设置按钮背景颜色
     *
     * @param normalColor
     * @param prssedClor
     */
    fun setBackgroundColor(normalColor: Int, pressedColor: Int) {

        mBgNormalColor = normalColor
        mBgPressedColor = pressedColor
        buildDraweableState()

    }

    /**
     * 设置按钮文字颜色
     *
     * @param normalColor
     * @param pressedColor
     */
    fun setTextColor(normalColor: Int, pressedColor: Int) {
        mTextPressedColor = pressedColor
        mTextNormalColor = normalColor
        buildColorDrawableState()

    }

    /**
     * 设置按钮背景颜色
     *
     * @param color
     */
    override fun setBackgroundColor(color: Int) {

        mBgNormalColor = color
        mBgPressedColor = color
        buildDraweableState()

    }

    /**
     * 设置按钮文字颜色
     *
     * @param color
     */
    override fun setTextColor(color: Int) {
        mTextPressedColor = color
        mTextNormalColor = color
        buildColorDrawableState()

    }

    companion object {
        var mNormalState = intArrayOf()
        var mPressState = intArrayOf(android.R.attr.state_pressed, android.R.attr.state_enabled)
    }
}